Implement "Use last" lint

HarrisonMc555 opened the pull request:
Closes #3673 

This lint checks the use of `x.get(x.len() - 1)` and suggests `x.last()` (where `x` is a `Vec`).

There's at least one more thing that needs to happen here. I believe I am correctly checking most of the scenarios and avoiding false positives. However, if different `Vec`s are used for the `x.get` and `y.len`, then it will emit a false positive. In other words, I need to check to make sure the `Vec`s are the same.

Also, a lot of these commits were temporary and not helpful to the project history...am I supposed to squash on merge? If so, how do I do that?

changelog: New lint: `get_last_with_len`

oli-obk commented on clippy_lints/src/lib.rs:
```diff
@@ -6,7 +6,8 @@
 #![feature(range_contains)]
 #![allow(clippy::missing_docs_in_private_items)]
-#![recursion_limit = "256"]
+// #![recursion_limit = "256"]
```
remove this comment

oli-obk commented on clippy_lints/src/use_last.rs:
```diff
@@ -0,0 +1,123 @@
+            // let _ = println!("LHS of sub method has an arg");
+
+            // TODO: Is this a valid way to check if they reference the same vector?
+            // if arg_lhs_struct.hir_id == struct_calling_on.hir_id;
```
you need to check whether the two expressions are equal by structure. I think we have an `expr_eq` function or similar code somewhere in clippy.

oli-obk commented on tests/ui/use_last.rs:
```diff
@@ -0,0 +1,36 @@
+fn dont_use_last() -> Option<i32> {
+    let x = vec![2, 3, 5];
+    let last_element = x.get(x.len() - 1); // ~ERROR Use x.last()
+    last_element.map(|val| val + 1) // To avoid warnings
```
you could just use `let _ = x.get ....`. The underscore silences the warnings, too

phansch commented:
friendly ping @HarrisonMc555 would you like to continue with this? Feel free to ask questions if you're stuck somewhere.

bors commented:
:umbrella: The latest upstream changes (presumably #3848) made this pull request unmergeable. Please resolve the merge conflicts.

HarrisonMc555 commented:
Yes, sorry. I've been caught up with other things. I'll try to get this done within the next week or two. Thanks for the ping!

HarrisonMc555 commented:
I believe I finished everything, sorry it took a while. I can squash if you want.

Also, I couldn't run `rustfmt`. When I run `cargo fmt` I get:

```text
error: 'cargo-fmt' is not installed for the toolchain 'master'
```

When I run `rustup component add rustfmt`, I get:

```
error: toolchain 'master' does not support components
```

I looked for an issue and I saw a merged pull request about switching to nightly (#3586) but I couldn't get that to work either. Can we add something to [DOCUMENTING.md](https://github.com/rust-lang/rust-clippy/blob/dc825a36f234984342a183d03a12d0263f8b72b3/doc/adding_lints.md#Running-rustfmt)?

bors commented:
:umbrella: The latest upstream changes (presumably #3968) made this pull request unmergeable. Please resolve the merge conflicts.

HarrisonMc555 commented:
I was wondering why the CI build failed...I forgot how fast things can change :-)

I just merged the changes, updated the README, and ran `rustfmt` (actually I ran rustfmt afterwards but it didn't change anything, thank goodness).

I think we should be good, we'll see in a few minutes if the CI build passes now. Thanks for the help everyone!

HarrisonMc555 commented:
So Travis CI failed, and I'm not sure why. I just re-ran `git pull upstream master` into my `master` branch and `git merge master` into my `use_last` branch.

I then ran `cargo test`, and all the tests pass. Any tips on resolving this? I'm not sure what's going on.

andrehjr commented:
@HarrisonMc555 Seems like it's still failing because of `cargo fmt` 🤔 maybe you need to update rustfmt? Anyway, if you fix the following, you should be good to go.

```
Diff in /Users/travis/build/rust-lang/rust-clippy/clippy_lints/src/use_last.rs at line 1:
 //! lint on using `x.get(x.len() - 1)` instead of `x.last()`
 
-use crate::utils::{match_type, paths, span_lint_and_sugg,
-                   snippet_with_applicability, SpanlessEq};
+use crate::utils::{match_type, paths, snippet_with_applicability, span_lint_and_sugg, SpanlessEq};
+use if_chain::if_chain;
+use rustc::hir::{Expr, ExprKind};
 use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
 use rustc::{declare_tool_lint, lint_array};
-use rustc::hir::{Expr, ExprKind};
 use rustc_errors::Applicability;
-use syntax::ast::{LitKind};
-use if_chain::if_chain;
+use syntax::ast::LitKind;
 
 /// **What it does:** Checks for using `x.get(x.len() - 1)` instead of `x.last()`.
 ///
The command "if [ -z ${INTEGRATION} ]; then```

andrehjr commented on clippy_lints/src/use_last.rs:
```diff
@@ -0,0 +1,107 @@
+    fn name(&self) -> &'static str {
+        "UseLast"
+    }
+}
```
You should be able to use `declare_lint_pass!` here. [See](https://github.com/rust-lang/rust-clippy/issues/3917) and also [here](https://github.com/rust-lang/rust-clippy/pull/3968)

HarrisonMc555 commented:
Ok I finally figured out how to run rustfmt. This is how I did it:

```shell
rustup component add --toolchain=nightly rustfmt
rustup run nightly rustfmt path/to/file.rs
```

Also I used the new `declare_lint_pass!` macro. Hopefully this passes CI now :-)

HarrisonMc555 commented:
@andrehjr everything passed now, looks like we should be able to merge. Let me know if there's anything else that needs to be done!

HarrisonMc555 commented:
Hmm so I had a passing CI build, but some other commits were pushed before this was merged. I pulled the new commits and everything merged on my machine fine.

Now, the Linux CI build passes but the macOS build fails. How do I go about fixing this? I don't have a macOS machine available.

flip1995 commented on tests/ui/use_last.rs:
```diff
@@ -0,0 +1,36 @@
+    last_element + 1 // To avoid warnings
+}
+
+// False positive test (currently failing)
```
This is probably not relevant anymore?

Also: https://github.com/rust-lang/rust-clippy/pull/3832/files#r262524555

phansch commented:
> Now, the Linux CI build passes but the macOS build fails. 

Woops, I forgot to submit the comment yesterday. It was only a spurious failure and a restart of the Travis build fixed it.

flip1995 commented on clippy_lints/src/use_last.rs:
```diff
@@ -0,0 +1,96 @@
+            if let rustc::hir::ExprKind::Binary(ref op, ref lhs, ref rhs) = get_index_arg.node;
+
+            // Binary operation is a subtraction
+            if op.node == rustc::hir::BinOpKind::Sub;
```
You can import `use rustc::hir::BinOpKind` and already match on this in the `if let` above.

HarrisonMc555 replied:
The only way I could find to do this was to add a struct pattern matching inside. Is that what you wanted?

flip1995 commented on clippy_lints/src/use_last.rs:
```diff
@@ -0,0 +1,96 @@
+
+            // Argument to "get" is a binary operation
+            if let Some(get_index_arg) = args.get(1);
+            if let rustc::hir::ExprKind::Binary(ref op, ref lhs, ref rhs) = get_index_arg.node;
```
```suggestion
            if let rustc::hir::ExprKind::Binary(BinOpKind::Sub, lhs, rhs) = &get_index_arg.node;
```

You should be able to remove these `ref`s by borrowing the `node`

HarrisonMc555 replied:
I think the syntax changed and now it requires a full struct (see above).

flip1995 commented on clippy_lints/src/use_last.rs:
```diff
@@ -0,0 +1,96 @@
+            if op.node == rustc::hir::BinOpKind::Sub;
+
+            // LHS of subtraction is "x.len()"
+            if let ExprKind::MethodCall(ref arg_lhs_path, _, ref lhs_args) = lhs.node;
```
```suggestion
            if let ExprKind::MethodCall(arg_lhs_path, _, lhs_args) = &lhs.node;
```

flip1995 commented on clippy_lints/src/use_last.rs:
```diff
@@ -0,0 +1,96 @@
+            if SpanlessEq::new(cx).eq_expr(struct_calling_on, arg_lhs_struct);
+
+            // RHS of subtraction is 1
+            if let ExprKind::Lit(ref rhs_lit) = rhs.node;
```
```suggestion
            if let ExprKind::Lit(rhs_lit) = &rhs.node;
```

flip1995 commented on clippy_lints/src/use_last.rs:
```diff
@@ -0,0 +1,96 @@
+
+            let mut applicability = Applicability::MachineApplicable;
+            let vec_name = snippet_with_applicability(
+                cx, struct_calling_on.span, "x", &mut applicability);
```
Can you put this in the `then` block (and the `let mut appl...`)? 

Also formatting on the `snippet_with_applicability`

HarrisonMc555 replied:
What formatting were you referring to? I put as much on one line as fit in 100 characters. Is that what we go for in clippy?

flip1995 commented on clippy_lints/src/use_last.rs:
````diff
@@ -0,0 +1,96 @@
+    /// let x = vec![2, 3, 5];
+    /// let last_element = x.last();
+    /// ```
+    pub USE_LAST,
````
We need another name here. Writing `allow(clippy::use_last)` implies that you want to allow `x.last()`. Something like `GET_LAST_WITH_LEN` maybe. You can find lint naming conventions here: #2845. Feel free to chose another name, if you come up with something better.

HarrisonMc555 replied:
I like `get_last_with_len`. All the files should be changed now.

bors commented:
:umbrella: The latest upstream changes (presumably #4039) made this pull request unmergeable. Please resolve the merge conflicts.

HarrisonMc555 commented:
I have some more time this week, so I should get it done within the next day or two. However, I end up working on a computer both at school and at home, so I'm pushing and pulling to keep them in sync...which I believe will trigger Travis builds, even when I don't intend it to. Is there any way to stop this? I just feel guilty causing an unnecessary build takes time away from necessary builds.

flip1995 commented:
You could push to a branch unrelated to this PR and when you're finished merge that branch into the branch of this PR.

HarrisonMc555 commented:
Ok, I think the pull request is ready. All the tests pass, and I believe I've addressed all the concerns. I've left the "conversations" open that I didn't take as literal suggestions. Sorry for not using the new GitHub built-in suggestion inserter thing.

flip1995 commented on clippy_lints/src/get_last_with_len.rs:
```diff
@@ -0,0 +1,98 @@
+            if let LitKind::Int(rhs_value, ..) = rhs_lit.node;
+            if rhs_value == 1;
+
+
```
Multiple empty lines, remove these.

HarrisonMc555 replied:
I assume one empty line is ok?

flip1995 commented on clippy_lints/src/get_last_with_len.rs:
```diff
@@ -0,0 +1,98 @@
+
+            then {
+                let mut applicability = Applicability::MachineApplicable;
+                let vec_name = snippet_with_applicability(cx, struct_calling_on.span, "vec",
```
Formatting:
```rust
func(
    arg1,
    arg2,
    ...
);
```

flip1995 commented on clippy_lints/src/get_last_with_len.rs:
```diff
@@ -0,0 +1,98 @@
+                    cx,
+                    GET_LAST_WITH_LEN,
+                    expr.span,
+                    &format!("Use `{}.last()` instead of `{}.get({}.len() - 1)`", vec_name,
```
```suggestion
                    &format!("Use `{0}.last()` instead of `{0}.get({0}.len() - 1)`", vec_name),
```

flip1995 replied:
Oh and the error message shouldn't be the suggestion. Instead of writing "Do this instead of that", you should write "You did this". The suggestion how to fix this is below.

In this case you could write: "accessing last element of vector with `{0}.get({0}.len() - 1)`".

FYI: error messages in rust always start lowercase and don't have punctuation (at the end).

HarrisonMc555 replied:
Thanks for the input!

flip1995 commented on tests/ui/get_last_with_len.stderr:
```diff
@@ -0,0 +1,10 @@
+error: unknown clippy lint: clippy::get_last_with_len
+  --> $DIR/get_last_with_len.rs:1:9
+   |
+LL | #![warn(clippy::get_last_with_len)]
```
You have to `register_late_pass(box UseLast)` in `lib.rs`. See https://github.com/rust-lang/rust-clippy/blob/master/doc/adding_lints.md#lint-declaration

HarrisonMc555 replied:
Thanks. I always forget about that, since it's the only thing that's not automated.

flip1995 commented on clippy_lints/src/get_last_with_len.rs:
```diff
@@ -0,0 +1,100 @@
+                let vec_name = snippet_with_applicability(
+                    cx,
+                    struct_calling_on.span, "vec",
+                    &mut applicability);
```
```suggestion
                    &mut applicability,
                );
```

flip1995 commented on clippy_lints/src/get_last_with_len.rs:
```diff
@@ -0,0 +1,100 @@
+                    cx,
+                    GET_LAST_WITH_LEN,
+                    expr.span,
+                    &format!("accessing last element with `{}.get({}.len() - 1)`",
```
```suggestion
                    &format!("accessing last element with `{0}.get({0}.len() - 1)`", vec_name),
```

bors commented:
:umbrella: The latest upstream changes (presumably #4088) made this pull request unmergeable. Please resolve the merge conflicts.

Manishearth commented:
Can you rebase over master instead of merging?

Manishearth commented:
It may be easier to create a fresh commit with the diff from the PR

HarrisonMc555 commented:
@Manishearth sure thing. What is the recommended way to do that? Do I create a patch with all the changes and add it to a new branch or delete the old commits from this branch or what?

flip1995 commented:
Since there are currently now conflicts with the master branch, the easiest thing to do would be this https://github.com/rust-lang/rust-clippy/pull/3832#pullrequestreview-210711596

That squashes commits, which you then can force push (`--force-with-lease`) to this branch.

HarrisonMc555 commented:
Did I do this right? I started doing `git rebase -i` but there were like 37 steps and they seemed redundant. So I just created a patch from master, reset back to master, applied the patch, and did `git push --force-with-lease`. Let me know if that's the right idea or if I need to do the tedious `git rebase -i` stuff :smile:

Manishearth commented:
That's fine! That's what I meant by creating a fresh commit.

HarrisonMc555 commented:
Ok great! So is everything done now? I think GitHub is saying there's still a change requested from you @Manishearth.

flip1995 commented:
Nope, that's all. Thanks for all the work!

@bors r+

bors commented:
:pushpin: Commit 32714919df697ee1e31211021315815ff51f64f3 has been approved by `flip1995`

<!-- @bors r=flip1995 32714919df697ee1e31211021315815ff51f64f3 -->

bors commented:
:hourglass: Testing commit 32714919df697ee1e31211021315815ff51f64f3 with merge 4a2ff35e0ce571b2d02cb1f921fb7a2a7b755703...
<!-- homu: {"type":"BuildStarted","head_sha":"32714919df697ee1e31211021315815ff51f64f3","merge_sha":"4a2ff35e0ce571b2d02cb1f921fb7a2a7b755703"} -->

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request is currently being tested. If there's no response from the continuous integration service, you may use `retry` to trigger a build again.

bors commented:
:pushpin: Commit 32714919df697ee1e31211021315815ff51f64f3 has been approved by `flip1995`

<!-- @bors r=flip1995 32714919df697ee1e31211021315815ff51f64f3 -->

bors commented:
:hourglass: Testing commit 32714919df697ee1e31211021315815ff51f64f3 with merge 64f4f622af397fb686b27d64f9678b4dfdd66dcd...
<!-- homu: {"type":"BuildStarted","head_sha":"32714919df697ee1e31211021315815ff51f64f3","merge_sha":"64f4f622af397fb686b27d64f9678b4dfdd66dcd"} -->

bors commented:
:broken_heart: Test failed - [checks-travis](https://travis-ci.com/rust-lang/rust-clippy/builds/112553583)
<!-- homu: {"type":"BuildFailed","builder_url":"https://travis-ci.com/rust-lang/rust-clippy/builds/112553583","builder_name":"checks-travis"} -->

flip1995 commented:
https://travis-ci.com/rust-lang/rust-clippy/jobs/201732318#L1010

cc #4121

HarrisonMc555 commented:
There was a minor hiccup when attempting to merge formatting suggestions but it's been resolved. All tests are passing and all changes are approved. Are we ready to merge?

There have been a few commits since I squashed, do you want me to merge, re-squash, and update the branch again?

flip1995 commented:
> do you want me to merge, re-squash, and update the branch again?

No it should be good as is, so this should work:

@bors r+

(Sorry, I was busy the past 4 days)

bors commented:
:pushpin: Commit 42d849c1853d47d2b392d449077cea944b7861a8 has been approved by `flip1995`

<!-- @bors r=flip1995 42d849c1853d47d2b392d449077cea944b7861a8 -->
<!-- homu: {"type":"Approved","sha":"42d849c1853d47d2b392d449077cea944b7861a8","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit 42d849c1853d47d2b392d449077cea944b7861a8 with merge cf81a3b6d01cdacae0a391c9cfbd17b3ce801d78...
<!-- homu: {"type":"BuildStarted","head_sha":"42d849c1853d47d2b392d449077cea944b7861a8","merge_sha":"cf81a3b6d01cdacae0a391c9cfbd17b3ce801d78"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/rust-clippy/builds/113286488), [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/rust-clippy/builds/24834583)
Approved by: flip1995
Pushing cf81a3b6d01cdacae0a391c9cfbd17b3ce801d78 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"flip1995","base_ref":"master","builders":{"status-appveyor":"https://ci.appveyor.com/project/rust-lang-libs/rust-clippy/builds/24834583","checks-travis":"https://travis-ci.com/rust-lang/rust-clippy/builds/113286488"},"merge_sha":"cf81a3b6d01cdacae0a391c9cfbd17b3ce801d78"} -->

HarrisonMc555 commented:
Thanks @flip1995! No problem :-)
